lib/theme_check/shopify_liquid/source_manager.rb

```diff
@@ -1,81 +1,110 @@
-        else
-          download
+    class SourceManager
+      REQUIRED_FILE_NAMES = %i(filters objects tags latest).freeze
+
+      class << self
+        def download_or_refresh_files(destination = default_destination)
+          if has_required_files?(destination)
+            refresh(destination)
+          else
+            download(destination)
+          end
         end
-      end
 
-      def download
-        create_destination unless destination_exist?
+        def download(destination = default_destination)
+          Dir.mkdir(destination) unless destination.exist?
 
-        required_file_names.each do |file_name|
-          download_file(local_path(file_name), remote_path(file_name))
+          REQUIRED_FILE_NAMES.each do |file_name|
+            download_file(local_path(file_name, destination), remote_path(file_name))
+          end
         end
-      end
 
-      def local_path(file_name)
-        documentation_directory + "#{file_name}.json"
-      end
+        def refresh(destination = default_destination)
+          refresh_threads << Thread.new { refresh_thread(destination) }
+        end
 
-      private
+        def local_path(file_name, destination = default_destination)
+          destination + "#{file_name}.json"
+        end
 
-      DOCUMENTATION_FETCH_URL = "https://github.com/Shopify/theme-liquid-docs/raw/main/data"
-      DOCUMENTATION_DIRECTORY = Pathname.new("#{__dir__}/../../../data/shopify_liquid/documentation")
-      REQUIRED_FILE_NAMES = [:filters, :objects, :tags].freeze
+        def has_required_files?(destination = default_destination)
+          REQUIRED_FILE_NAMES.all? { |file_name| local_path(file_name, destination).exist? }
+        end
 
-      def remote_path(file_name)
-        "#{documentation_fetch_url}/#{file_name}.json"
-      end
+        def wait_downloads
+          refresh_threads.each(&:join)
+        end
 
-      def download_file(local, remote)
-        File.open(local, "wb") do |file|
-          content = open_uri(remote)
+        private
 
-          file.write(content)
+        def refresh_thread(destination)
+          return unless refresh_needed?(destination)
+
+          Dir.mktmpdir do |tmp_dir|
+            download(Pathname.new(tmp_dir))
+
+            FileUtils.cp_r("#{tmp_dir}/.", destination)
+
+            mark_all_indexes_outdated
+          end
         end
-      end
 
-      def open_uri(remote)
-        # require at usage point to not slow down theme-check on startup
-        require 'open-uri'
+        def refresh_needed?(destination)
+          local_latest_content = local_path(:latest, destination).read
+          remote_latest_content = open_uri(remote_path(:latest))
 
-        URI.parse(remote).open.read
-      end
+          revision(local_latest_content) != revision(remote_latest_content)
+        end
 
-      def destination_exist?
-        documentation_directory.exist?
-      end
+        def revision(json_content)
```

Comment:
I'm not sure about `!` conventions, but if this method can raise an error, we might rename it to `revision!` ? :D

Reply:
so im hesitant to do this because raising an error here isn't necessarily dangerous in theme-check? refresh erroring out will prevent doc updates but the user will still be able to use theme-check (and even code completion) 😄 

also for internal errors, we need to raise in order to send a message to the client that something went wrong (they're caught and sent [here](https://github.com/Shopify/theme-check/blob/main/lib/theme_check/language_server/server.rb#L142-L149) but the handler keeps chugging along 😎 )